httpcore/_async/http2.py

```diff
@@ -123,10 +123,21 @@ async def handle_async_request(self, request: Request) -> Response:
                 )
                 self._max_streams_semaphore = AsyncSemaphore(local_settings_max_streams)
 
-                for _ in range(local_settings_max_streams - self._max_streams):
-                    await self._max_streams_semaphore.acquire()
-
-        await self._max_streams_semaphore.acquire()
+                try:
+                    for _ in range(local_settings_max_streams - self._max_streams):
+                        await self._max_streams_semaphore.acquire()
+                except BaseException:
+                    with AsyncShieldCancellation():
+                        async with self._state_lock:
+                            await self._state_housekeeping()
+                    raise
+        try:
+            await self._max_streams_semaphore.acquire()
+        except BaseException:  # pragma: no cover
+            with AsyncShieldCancellation():
+                async with self._state_lock:
+                    await self._state_housekeeping()
+            raise
```

Comment:
Should `self._state_housekeeping` just be a context manager then you could simplify both of these cases to

```python
async with self._cleanup_on_error():
    ...  # do work
```

and the try/except logic can be handled in the context manager's yield.

I find this easier to read and it consolidates logic for the (slightly dangerous) capture of a `BaseException`.

Reply:
> I find this easier to read and it consolidates logic for the (slightly dangerous) capture of a BaseException.

We always catch the `BaseException` in `httpcore` for cleanup logic, so I believe it is safe.

> Should self._state_housekeeping just be a context manager then you could simplify both of these cases to

Can you provide an example of how it might look? Context managers are frequently used for try/finally logic rather than try/except logic, am I correct? Only when an exception occurs do we need to do housekeeping.

Reply:
Or do we really need that much extra complexity?

Reply:
I think `try/except` logic is common as well.

Here's a sketch of what I would do https://github.com/encode/httpcore/commit/178e7d1b8cd93a581ce23bea172bdea13a38d799

I find the intent clearer here — I'm not sure it's more complex than maintaining the repeated implementation for properly handling an exception with the cancellation shield and state lock.

Reply:
Let's keep it for another PR. `httpcore` now uses this type of cleanup everywhere, and if we want to replace common exception handling with context managers, we'll probably want to do it elsewhere as well.

Reply:
For example `_response_closed` call

```python
except BaseException as exc:
    with ShieldCancellation():
        with Trace("response_closed", logger, request) as trace:
            self._response_closed()
    raise exc
```

Reply:
I don't quite think it follows that we'd want to do it elsewhere, the example you gave here is specific to a closed response whereas the other one is "some statement, if interrupted, needs us to clean up the state".

Anyway we don't need to go back and forth here. It really depends what style @tomchristie prefers for the project.